solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java

```diff
@@ -17,54 +17,49 @@
 
 package org.apache.solr.util.circuitbreaker;
 
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Metric;
 import java.lang.invoke.MethodHandles;
-import java.lang.management.ManagementFactory;
-import java.lang.management.OperatingSystemMXBean;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.metrics.SolrMetricManager;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Tracks current CPU usage and triggers if the specified threshold is breached.
  *
- * <p>This circuit breaker gets the average CPU load over the last minute and uses that data to take
- * a decision. We depend on OperatingSystemMXBean which does not allow a configurable interval of
- * collection of data. //TODO: Use Codahale Meter to calculate the value locally.
- *
- * <p>The configuration to define which mode to use and the trigger threshold are defined in
- * solrconfig.xml
+ * <p>This circuit breaker gets the recent average CPU usage and uses that data to take a decision.
+ * We depend on OperatingSystemMXBean which does not allow a configurable interval of collection of
+ * data.
  */
 public class CPUCircuitBreaker extends CircuitBreaker {
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
-  private static final OperatingSystemMXBean operatingSystemMXBean =
-      ManagementFactory.getOperatingSystemMXBean();
 
+  private static boolean warningLogged = false;
```

Comment:
edge case: could the circuit breaker be used from multiple threads and so `boolean` here needs to be supportive of potentially concurrent access?

Reply:
The consequence of a race condition here would be that there would be more than one WARN log line before it gets muted, since we don't have code that resets it to false anywhere. Think we can live with that. Other concerns?

Reply:
So an alternative to `boolean` would be `AtomicBoolean` then, say. Since the variable is only read (and set) in an error scenario there then maybe wouldn't be any concern re: atomic being more expensive.

But actually, I wonder, if failure to acquire the inputs is systemic, wouldn't one want to know as early as possible e.g. in the constructor and to at least log a warning there, maybe an error or to throw an exception even?
* That's not specific to this class I guess.
* In the case of a search circuit breaker one could imagine ping-like behaviour resulting in the warning being logged ahead of 'real' traffic reaching the core.
* In the case of an update circuit breaker would it conceptually only be the first real update request that results in a warning and that could be confusing or undesirable, whereas a warning on startup is more harmless?

If the constructor made a `isTripped(true)` call then no `warningLogged` member variable would be needed and that would also help with the _the value of these parameters will be set correctly before invoking getDebugInfo()_ assumption at line 42/43.

```
public boolean isTripped() {
  return isTripped(false);
}

private boolean isTripped(boolean logWarning) {
  ...
}
```

Reply:
Here's a patch that I believe does the trick. It will log an ERROR duing initialization and disable the plugin. On every call it will log DEBUG message about it being disabled:
```patch
diff --git a/solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java b/solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java
--- a/solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java	(revision d2951a00f6427835fce82304e481a0536cc8f11c)
+++ b/solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java	(date 1693865423560)
@@ -20,6 +20,7 @@
 import com.codahale.metrics.Gauge;
 import com.codahale.metrics.Metric;
 import java.lang.invoke.MethodHandles;
+import org.apache.solr.common.util.NamedList;
 import org.apache.solr.core.SolrCore;
 import org.apache.solr.metrics.SolrMetricManager;
 import org.slf4j.Logger;
@@ -35,7 +36,7 @@
 public class CPUCircuitBreaker extends CircuitBreaker {
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
-  private static boolean warningLogged = false;
+  private static boolean enabled = true;
   private double cpuUsageThreshold;
   private final SolrCore core;
 
@@ -51,19 +52,31 @@
   }
 
   @Override
-  public boolean isTripped() {
-    double localAllowedCPUUsage = getCpuUsageThreshold();
+  public void init(NamedList<?> args) {
+    super.init(args);
     double localSeenCPUUsage = calculateLiveCPUUsage();
 
     if (localSeenCPUUsage < 0) {
-      if (!warningLogged && log.isWarnEnabled()) {
-        String msg = "Unable to get CPU usage";
-        log.warn(msg);
-        warningLogged = true;
+      String msg =
+          "Initialization failure for CPU circuit breaker. Unable to get 'systemCpuLoad', not supported by the JVM?";
+      if (log.isErrorEnabled()) {
+        log.error(msg);
       }
+      enabled = false;
+    }
+  }
 
+  @Override
+  public boolean isTripped() {
+    if (!enabled) {
+      if (log.isDebugEnabled()) {
+        log.debug(
+            "CPU circuit breaker is disabled due to initialization failure, will never trip.");
+      }
       return false;
     }
+    double localAllowedCPUUsage = getCpuUsageThreshold();
+    double localSeenCPUUsage = calculateLiveCPUUsage();
 
     allowedCPUUsage.set(localAllowedCPUUsage);

```

Reply:
You closed the comment without merging https://github.com/wrunderwood/solr/pull/1 and without commenting on whether you like the idea of disabling the CB on init failure as opposed to spamming the log with WARN msgs for every single request.

Reply:
I committed the proposal.